crates/shared/src/gas_price_estimation.rs

```diff
@@ -128,3 +132,76 @@ impl GasPriceEstimating for FakeGasPriceEstimator {
         Ok(*self.0.lock().unwrap())
     }
 }
+
+/// Gas estimator which will never reuse an old gas price estimate. This is useful to supply a
+/// cheap finger print for otherwise identical transactions.
+/// In case an estimate would be a duplicate the estimator increases the estimate until it finds
+/// the first unique value.
+/// To keep the unique prices as close to the expected prices as possible a
+/// `FingerPrintingGasEstimator` should be as short lived as possible.
+pub struct FingerPrintingGasPriceEstimator<'a> {
+    inner: &'a dyn GasPriceEstimating,
+    previous_estimates: Mutex<HashSet<u64>>,
+}
+
+impl<'a> FingerPrintingGasPriceEstimator<'a> {
+    pub fn new(inner: &'a dyn GasPriceEstimating) -> Self {
+        Self {
+            inner,
+            previous_estimates: Default::default(),
+        }
+    }
+
+    fn closest_unique_gas_estimate(&self, mut target: EstimatedGasPrice) -> EstimatedGasPrice {
+        let mut closest_unique_price = target.legacy;
```

Comment:
Hm, I would avoid using `legacy` whenever possible because of the: https://github.com/cowprotocol/services/issues/119
So, I here I would rely on the eip1559 type.